cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -0,0 +1,285 @@
+    return Status::OK();
+  }
+};
+
+template <typename Type>
+struct ResolveExec {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch.length == 0) return Status::OK();
+
+    if (batch[0].kind() == Datum::ARRAY) {
+      if (batch[1].kind() == Datum::ARRAY) {
+        if (batch[2].kind() == Datum::ARRAY) {  // AAA
+          return IfElseFunctor<Type>::Call(ctx, *batch[0].array(), *batch[1].array(),
+                                           *batch[2].array(), out->mutable_array());
+        } else {  // AAS
+          return IfElseFunctor<Type>::Call(ctx, *batch[0].array(), *batch[1].array(),
+                                           *batch[2].scalar(), out->mutable_array());
+        }
+      } else {
+        return Status::Invalid("");
+        //        if (batch[2].kind() == Datum::ARRAY) {  // ASA
+        //          return IfElseFunctor<Type, true>::Call(ctx, *batch[0].array(),
+        //                                                 *batch[2].array(),
+        //                                                 *batch[1].scalar(),
+        //                                                 out->mutable_array());
+        //        } else {  // ASS
+        //          return IfElseFunctor<Type>::Call(ctx, *batch[0].array(),
+        //          *batch[1].scalar(),
+        //                                           *batch[2].scalar(),
+        //                                           out->mutable_array());
+        //        }
+      }
+    } else {
+      if (batch[1].kind() == Datum::ARRAY) {
+        return Status::Invalid("");
+        //        if (batch[2].kind() == Datum::ARRAY) {  // SAA
+        //          return IfElseFunctor<Type>::Call(ctx, *batch[0].scalar(),
+        //          *batch[1].array(),
+        //                                           *batch[2].array(),
+        //                                           out->mutable_array());
+        //        } else {  // SAS
+        //          return IfElseFunctor<Type>::Call(ctx, *batch[0].scalar(),
+        //          *batch[1].array(),
+        //                                           *batch[2].scalar(),
+        //                                           out->mutable_array());
+        //        }
+      } else {
+        if (batch[2].kind() == Datum::ARRAY) {  // SSA
+          return Status::Invalid("");
+          //          return IfElseFunctor<Type>::Call(ctx, *batch[0].scalar(),
+          //          *batch[1].scalar(),
+          //                                           *batch[2].array(),
+          //                                           out->mutable_array());
+        } else {  // SSS
+          return IfElseFunctor<Type>::Call(ctx, *batch[0].scalar(), *batch[1].scalar(),
+                                           *batch[2].scalar(), out->scalar().get());
+        }
+      }
+    }
+  }
+};
+
+void AddPrimitiveKernels(const std::shared_ptr<ScalarFunction>& scalar_function,
```

Comment:
A lot of these function names are a little generic, can we rename them to reflect that they're for IfElse? We might end up putting other similar kernels here.